core/src/main/java/org/apache/druid/segment/data/VByte.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.data;
+
+import java.nio.ByteBuffer;
+
+public class VByte
+{
+  /**
+   * Read a variable byte (vbyte) encoded integer from a {@link ByteBuffer} at the current position.
+   *
+   * vbyte encoding stores values in the last 7 bits of a byte and reserves the high bit for the 'contination'. If 0,
+   * one or more aditional bytes must be read to complete the value, and a 1 indicates the terminal byte. Because of
+   * this, it can only store positive values, and larger integers can take up to 5 bytes.
+   *
+   * implementation based on:
+   * https://github.com/lemire/JavaFastPFOR/blob/master/src/main/java/me/lemire/integercompression/VariableByte.java
+   *
+   */
+  public static int readInt(ByteBuffer buffer)
```

Comment:
I hadn't got to playing around with this yet to make this change. It would probably be fine for the iterator to read this way, but the get method at least would benefit from this change

Reply:
One of the complications of this is that my current implementation of the `VByte` encoding which i'm using to store the ints for prefix and value lengths is a variable length itself, and so the current implementation is based on using the buffer to advance the position so the read methods can just return an `int` instead of a Pair of `int` and how long the int was so I know how far to advance the buffer, so will need to try to think of how to handle this without making even more extra junk, which I'm not sure is possible.

One garbage per block of values seems like maybe less than 2+ garbage per value, so maybe need to think a bit more about this.

Reply:
i did manage to pretty heavily reduce garbage between these comments and the latest version of the codebase, in these commits specifically i think https://github.com/apache/druid/pull/12277/files/20bdc3eb622fd46729eff785072645a8cd16b633..0e8cad11d6ad4ef5081b0a86dbbd0d9ea5289e08, so I think its in an ok place, I just forgot to update this thread